### PR TITLE
Unit and Integration Testing

### PR DESCRIPTION
**Pull request by AndresSan6**

### CODE REVIEW

File: library/src/main/java/com/skytouch/springboot/view/product/Product.java [hunk 1]
```diff
@@ -9,33 +11,45 @@
...
-    static final String PRICE = "price";
-    static final String DESCRIPTION = "description";
+public class Product {
 
     @Id
     @GeneratedValue(strategy = GenerationType.AUTO)
     private Long id;
 
+    @NotBlank
     private String name;
```
Comment (ragore09) [inline library/src/main/java/com/skytouch/springboot/view/product/Product.java]: Can name and description be `null`?

File: application/pom.xml [hunk 2]
```diff
@@ -39,10 +46,16 @@
...
-            <scope>compile</scope>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-test</artifactId>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>org.easymock</groupId>
+            <artifactId>easymockclassextension</artifactId>
+            <version>3.1</version>
```
Comment (fjavierponce) [inline application/pom.xml]: are you using dependency management? better if you manage your versions in one.

File: library/src/main/java/com/skytouch/springboot/view/product/ProductConstants.java [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package com.skytouch.springboot.view.product;
+
+public interface ProductConstants {
```
Comment (ragore09) [inline library/src/main/java/com/skytouch/springboot/view/product/ProductConstants.java]: why did you decide to use an interface instead of an enum?

File: microservice/src/test/java/com/skytouch/springboot/view/consumer/ConsumerTest.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+import org.junit.Before;
+import org.junit.Test;
+
+public class ConsumerTest {
+
+    @Before
+    public void setUp() {
+    }
+
+    @Test
```
Comment (ragore09) [inline microservice/src/test/java/com/skytouch/springboot/view/consumer/ConsumerTest.java]: if these tests are still not implemented, you should not add the `@Test` annotation

File: application/src/test/java/com/skytouch/springboot/view/controller/ProductControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+    @After
+    public void tearDown() {
+        reset(producer);
+    }
+
+    @Test
+    public void listProducts() {
+        // Given
+        List<Product> list = new ArrayList<>();
```
Comment (fjavierponce) [inline application/src/test/java/com/skytouch/springboot/view/controller/ProductControllerTest.java]: what is the difference if you use Collections.emptyList()? you could also use a non-empty one and verify with containsAll(list). Looks good, just consider that.

File: application/src/test/java/com/skytouch/springboot/view/controller/ProductControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    @Test
+    public void productForm() {
+        // Given
+        ModelAndView modelAndView = productController.productForm();
+        // Then
+        assertThat(modelAndView).isNotNull();
+        assertThat(modelAndView.getViewName()).isEqualTo("productHome");
+        assertThat(modelAndView.getModelMap()).isNotEmpty();
+    }
+}
```
Comment (fjavierponce) [inline application/src/test/java/com/skytouch/springboot/view/controller/ProductControllerTest.java]: nitpick: it is a good practice to add a new line in the end of file https://thoughtbot.com/blog/no-newline-at-end-of-file

File: microservice/pom.xml [hunk 7]
```diff
@@ -28,6 +28,19 @@
...
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-test</artifactId>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>org.easymock</groupId>
+            <artifactId>easymockclassextension</artifactId>
+            <version>3.1</version>
```
Comment (fjavierponce) [inline microservice/pom.xml]: same for dependency management. A good example here, what if you need to change easymock version? you will need to change it everywhere you used it in the project.

File: application/pom.xml [hunk 8]
```diff
@@ -16,6 +16,13 @@
     </parent>
 
     <dependencies>
+        <dependency>
+            <groupId>com.skytouch.app</groupId>
+            <artifactId>library</artifactId>
+            <version>0.0.1-SNAPSHOT</version>
```
Comment (Joxebus) [inline application/pom.xml]: Better use a section on the top to set the versions and then you can use that properties here. https://maven.apache.org/pom.html#Properties

File: microservice/src/test/resources/application-test.yml [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    generate-ddl: true
+    hibernate:
+      ddl-auto: update
+      dialect: org.hibernate.dialect.MySQL5Dialect
+  datasource:
+    url: jdbc:mysql://127.0.0.1:3310/SkyTouchAppTest
+    username: root
+    password: ${DOCKER_MYSQL_PW}
+    driver-class-name: com.mysql.cj.jdbc.Driver
+#  rabbitmq:
```
Comment (Joxebus) [inline microservice/src/test/resources/application-test.yml]: If you are not using this configurations, delete them instead of comment.

File: application/pom.xml [hunk 10]
```diff
@@ -39,10 +46,16 @@
...
-            <artifactId>library</artifactId>
-            <version>0.0.1-SNAPSHOT</version>
-            <scope>compile</scope>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-test</artifactId>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>org.easymock</groupId>
```
[the remark this answers is not in the document]
Comment (AndresSan6) [reply application/pom.xml]: When started writing my tests, found EasyMock useful to mock what I tried to implement. Doing a research between EasyMock and Mockito, found that Mockito is considered as the best framework for mock in the Java community because for it easy and simple API. Will read about how to mock with Mockito and try to migrate to Mockito.

File: library/src/main/java/com/skytouch/springboot/view/product/ProductConstants.java (hunk 3 above)
Comment (AndresSan6) [reply library/src/main/java/com/skytouch/springboot/view/product/ProductConstants.java]: Reading about interfaces and enums, didn't knew that enum are exclusive for constants. Will make this change right away in the next commit.

File: application/src/test/java/com/skytouch/springboot/view/controller/ProductControllerTest.java (hunk 5 above)
Comment (AndresSan6) [reply application/src/test/java/com/skytouch/springboot/view/controller/ProductControllerTest.java]: I didn't understand the question. 

Reading about `Collections.emptyList()`, it returns a immutable list to which you can't add new elements, and I am creating in my test a non-empty list and adding elements to it, and asserting that this list I created contains the same product that the list my controller returns.
